TrackletAlgorithm/MemoryTemplateBinnedCM.h

```diff
@@ -27,8 +26,23 @@ template<class DataType, unsigned int NBIT_BX, unsigned int NBIT_ADDR, unsigned
 // (1<<NBIT_ADDR): depth of the memory for each BX
 // NBIT_BIN: number of bits used for binning; (1<<NBIT_BIN): number of bins
 // NCOPY: is the number of memory copies - same as number of TE used
+
+// N.B. The CMSSW version of this pretends to have NCOPY copies to anyone 
+// using this class, but doesn't actually instantiate them, to save memory.
+
 class MemoryTemplateBinnedCM{
 
+ private:
+
+#ifdef CMSSW_GIT_HASH
+  static constexpr bool isCMSSW = true;
+  static constexpr unsigned int NBIT_BX = 0;
+  static constexpr unsigned int NCP = 1;
+#else
+  static constexpr bool isCMSSW = false;
+  static constexpr unsigned int NCP = NCOPY;
```

Comment:
@tomalin Since you define `NCP` here, I would suggest using it throughout instead of `NCOPY`. In particular, here:
https://github.com/cms-L1TK/firmware-hls/blob/2d0110f714d902afe1735422c3dd9066216da94d/TrackletAlgorithm/MemoryTemplateBinnedCM.h#L74
and here:
https://github.com/cms-L1TK/firmware-hls/blob/2d0110f714d902afe1735422c3dd9066216da94d/TrackletAlgorithm/MemoryTemplateBinnedCM.h#L142

Reply:
To play safe, I want the CMSSW version of this class to look identical to the outside World to the non-CMSSW version. If function getNCopy() returned NCP, which is 1 when running in CMSSW and NCOPY when running outside CMSSW, this requirement would not be met. The danger is that other HLS modules might call getNCopy when running in CMSSW, and behave strangely because they get the answer 1. e.g. If downstream modules use getNCopy to determine the number of parallel modules that can process the data from the memory, then this number might be set to 1 in CMSSW, leading to greater truncation.

Reply:
Thank you for explaining. I understand now.